### ISSUE TITLE
可以用不同的虚拟机，运行同一个mpc文件吗

### ISSUE THREAD

**Issue by shazi4399** [FudanMPL/Garnet#62]
我想请教一个问题，我注意到在[https://github.com/FudanMPL/Garnet/blob/main/docs/xgboost-training.md](https://github.com/FudanMPL/Garnet/blob/main/docs/xgboost-training.md)，
给出了三个虚拟机（semi2k-with-conversion、rss-with-conversion、mal-rss-with-conversion）例子，

这三个虚拟机（或者称为协议）对应不同的场景需求，都可以执行xgboost.mpc描述的算法。

我想请问，这里的虚拟机可以是其他协议吗，比如说可以换成 fss-ring-party.x、mascot-party.x、replicated-ring-party.x等虚拟机，去执行 xgboost.mpc 文件吗？

感谢回复！

**Comment by GuopengLin** [comment]
目前来说，xgboost训练只指支持文档中描述的三个虚拟机。因为xgboost训练用到了我们所提出的一个环环转化协议，目前该协议的实现只在这三个虚拟机中实现。
具体来说，semi2k-with-conversion、rss-with-conversion、mal-rss-with-conversion是在原来的semi2k-party, replicated-ring-party.x, malicious-rep-ring-party.x这三个虚拟机上加了环环转化协议的优化。

如果一定要在其他虚拟机上运行的话，可以修改Complier/xgboost.py中的Gain函数，修改成如下（删除了环环转化的优化）：
    def Gain(self, g, y, y_pred, debug=False):
        assert len(g) == len(y)
        gradients = SquareLoss.gradient(y, y_pred)
        hessians = SquareLoss.hessian(y, y_pred)
        G = GroupSum(g, gradients)
        H = GroupSum(g, hessians)
        G_l = GroupPrefixSum(g, gradients)
        H_l = GroupPrefixSum(g, hessians)
        G_r = G - G_l
        H_r = H - H_l
        G_l_128 = G_l
        H_l_128 = H_l
        G_r_128 = G_r
        H_r_128 = H_r
        G_l_128_square = G_l_128 * G_l_128
        G_r_128_square = G_r_128 * G_r_128
        res = G_l_128_square / (H_l_128 + self.lamb) + G_r_128_square / (H_r_128 + self.lamb)
        return res
并且将编译命令修改为: python ./compile.py -R 128 xgboost即可。

这样做的缺点是，失去了环环转化优化，使得整个训练过程需要在128位的环上进行（小数据集64位的环应该也可以）。

**Comment by shazi4399** [comment]
明白了，非常感谢！
